app/src/main/java/org/merakilearn/ProfileActivity.kt

```diff
@@ -53,6 +63,88 @@ class ProfileActivity : AppCompatActivity() {
 
     }
 
+    private fun initSavedFile() {
+        val adapter = SavedFileAdapter {
+            showPopupMenu(it)
+        }
+        val padding = resources.getDimensionPixelSize(R.dimen.spacing_2x)
+        mBinding.recyclerview.layoutManager = GridLayoutManager(this, 2)
+        mBinding.recyclerview.addItemDecoration(GridSpacingDecorator(padding, padding, 2))
+        mBinding.recyclerview.adapter = adapter
+        viewModel.fetchSavedFile.observe(this, Observer {
+            updateAdapter(it)
+        })
+
+        mBinding.tvViewAll.setOnClickListener {
+            adapter.submitList(filesList)
+        }
+    }
+
+    private fun updateAdapter(
+        listData: List<Pair<String, String>>
+    ) {
+        val adapter = mBinding.recyclerview.adapter as SavedFileAdapter
+        filesList.clear()
+        filesList.addAll(listData)
+        if (listData.isEmpty()) {
+            mBinding.rlSavedFile.visibility = View.GONE
+        } else {
+            mBinding.rlSavedFile.visibility = View.VISIBLE
+        }
+        if (listData.isNotEmpty() && listData.size >= 4)
+            adapter.submitList(listData.subList(0, 4))
+        else
+            adapter.submitList(listData)
+    }
+
+    private fun showPopupMenu(it: Triple<String, String, View>) {
+        val popup = PopupMenu(this, it.third)
+        popup.menuInflater.inflate(R.menu.popup_menu_saved_file, popup.menu)
+        //registering popup with OnMenuItemClickListener
+        popup.setOnMenuItemClickListener { item ->
+            when (item.itemId) {
+                R.id.share -> shareFile(it.first)
+                R.id.delete -> deleteSavedFile(it.first)
+                R.id.copy -> copyContent(it.first)
+            }
+            true
+        }
+        popup.show()
+
+    }
+
+    private fun deleteSavedFile(first: String) {
+        viewModel.deleteFile(first).observe(this, Observer {
+            if (it) {
+               val newList= filesList.filter {pair->
+                    pair.first != first
+                }.toMutableList()
+                updateAdapter(newList)
+                filesList.clear()
+                filesList.addAll(newList)
+                toast(getString(R.string.file_deleted))
+            }
+        })
+    }
+
+    private fun copyContent(first: String) {
+        val code = File(first).bufferedReader().readLine()
+        if (!code.isNullOrBlank())
+            merakiNavigator.openPlayground(this, code)
+    }
+
+    private fun shareFile(first: String) {
+        val code = File(first).bufferedReader().readLine()
+        val sendIntent: Intent = Intent().apply {
+            action = Intent.ACTION_SEND
+            putExtra(Intent.EXTRA_TEXT, code)
+            type = "text/plain"
+        }
+        val shareIntent = Intent.createChooser(sendIntent, getString(org.navgurukul.playground.R.string.share_code))
+        startActivity(shareIntent)
```

Comment:
maybe we can add a check here to send file in case of big file, else we can share text

Reply:
File is saved with .py extension. Very less chance is there for any app other than Meraki which could handle that. Hence sharing text seems to be okay here. 
Also since this would be text file, file size can't be more than an MB for basic learning of python.

Reply:
Please suggest if something else can be done.